Add support for target_os = android

flxo opened the pull request:
Android supports memfds. Add `target_os = android` to the conditional compilation flags. Replace the use of raw libc::syscall` with a more generic `libc::fcntl` calls.

Test are fine on a AOSP Android 9.0:

```sh
~/memfd-rs ‹pr-android*› cargo test --target aarch64-linux-android                                     
   Compiling memfd v0.4.1-alpha.0 (/home/felix/memfd-rs)
    Finished test [unoptimized + debuginfo] target(s) in 0.49s
     Running unittests (target/aarch64-linux-android/debug/deps/memfd-9f7fbfff9cfff8d3)
remount succeeded
+ adb push /home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/memfd-9f7fbfff9cfff8d3 /data
/home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/memfd-9f7fbfff9cfff8d3: 1 file pushed. 21.4 MB/s (5790168 bytes in 0.258s)
+ adb shell /data/memfd-9f7fbfff9cfff8d3

running 1 test
test errors::error_send_sync ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ adb shell rm /data/memfd-9f7fbfff9cfff8d3
     Running tests/memfd.rs (target/aarch64-linux-android/debug/deps/memfd-18901d5b48e24950)
remount succeeded
+ adb push /home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/memfd-18901d5b48e24950 /data
/home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/memfd-18901d5b48e24950: 1 file pushed. 21.6 MB/s (6180424 bytes in 0.273s)
+ adb shell /data/memfd-18901d5b48e24950

running 4 tests
test test_memfd_default ... ok
test test_memfd_from_into ... ok
test test_memfd_multi ... ok
test test_memfd_no_cloexec ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

+ adb shell rm /data/memfd-18901d5b48e24950
     Running tests/sealing.rs (target/aarch64-linux-android/debug/deps/sealing-a99958c7b2daecde)
remount succeeded
+ adb push /home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/sealing-a99958c7b2daecde /data
/home/felix/memfd-rs/target/aarch64-linux-android/debug/deps/sealing-a99958c7b2daecde: 1 file pushed. 21.2 MB/s (6377448 bytes in 0.287s)
+ adb shell /data/sealing-a99958c7b2daecde

running 4 tests
test test_sealing_default ... ok
test test_sealing_add ... ok
test test_sealing_unsealed ... ok
test test_sealing_resize ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ adb shell rm /data/sealing-a99958c7b2daecde
```

nagisa commented on src/memfd.rs:
```diff
@@ -210,7 +210,7 @@ impl Memfd {
         let flags = sealing::seals_to_bitflags(seals);
         // UNSAFE(lucab): required syscall.
-        let r = unsafe { libc::syscall(libc::SYS_fcntl, fd, libc::F_ADD_SEALS, flags) };
+        let r = unsafe { libc::fcntl(fd, libc::F_ADD_SEALS, flags) };
```
I recall there being a reason to use the `syscall` interface. Not sure what it was, however.

flxo replied:
Ok. Do you remember? Using `libc::fcntl` feel more portable than the explicit sys call name. I replaced `syscall` with `fnctl` because I had build errors for `aarch64-linux-android` with the `SYS_fcntl` arg which is not exposed by libc. I haven't digged which syscall name to use instead which would also be an option with a `cfg(target_os = ...` guard.

lucab replied:
It looks like I wrote it that way initially, but I don't remember exactly why. Trying to retro-guess myself, it's likely one of those reasons:
 * `fcntl()` is a muxed wrapper in most libc implementations, and I didn't see a real reason to go through an additional opaque layer (and to make writing seccomp filters easier)
 * I was possibly playing with this on a platform where libc-rs was missing the function binding (uclibc? musl?) and so I went for the syscall directly
 * I was playing with the idea of making this work just with [sc](https://docs.rs/sc) and without libc-rs at all, but I never walked that path till the end

Or possibly a mix of all of them. The only interesting one nowadays is the first, but portability is likely a better argument.

flxo replied:
@lucab Thanks for your reply. Which way would you like to go? I can try to refactor this patch to use `libc::syscall` with`target_os` (and maybe `target_arch`) specific guards. Otherwise we can also check which targets are affected at all.

lucab replied:
@flxo if you are not in a hurry, it may be worth to just dig a bit more into libc-rs. If it is just missing a few `SYS_fcntl` definitions, it would be good to add it there; they have a self-service release process so it shouldn't add to much delay before being able to consume it here.
Instead, if the reason turns out to be more complex than that and we don't find any blocker for switching to the libc muxed wrapper, we can freely do that as a fallback.
In any case, thanks for looking into this :)

flxo replied:
@lucab I'll have a look why my first attempt with the direct use of `libc::syscall` didn't work for `aarch64-linux-android`.

flxo replied:
@lucab Turned out that `SYS_fcntl` is missing in `aarch64-linux-android` in libc. I opened [#2308](https://github.com/rust-lang/libc/pull/2308)

The flags `F_ADD_SEALS` and `F_GET_SEALS` are already included in libc for this target.
I'd propose to wait for the PR unless you're fine with the use of `libc::fcntl`...

flxo committed:
Fix build for target_os android

Add `target_os = android` to the conditional compilation flags.
Add targets `aarch64-linux-android` and `arm-linux-androideabi` with the
use of `cross` the the GH action.

flxo commented:
@lucab Hi. I'm back from vacation :-)

Looks like the merge of the `libc` patch did the trick.

lucab commented:
@flxo thanks, patch looks fine. It would be good to add an explicit version to the `libc` dependency in cargo manifest to make sure consumers are meeting the minimum requirement.

flxo commented:
@lucab Many thanks! Would you mind publishing a 0.4.1? There doesn't seem to be anything in the queue and I'd love to get rid of another `git` dependency.
